kzg_prover/bin/gen_commit_and_proofs.rs

```diff
@@ -0,0 +1,230 @@
+    halo2curves::{
+        bn256::{Bn256, Fr as Fp, G2Affine},
+        group::Curve,
+    },
+    poly::kzg::commitment::KZGCommitmentScheme,
+};
+use num_bigint::BigUint;
+use serde::{Deserialize, Serialize};
+use serde_json::to_string_pretty;
+use std::{fs::File, io::Write};
+use summa_solvency::{
+    circuits::{
+        univariate_grand_sum::UnivariateGrandSum,
+        utils::{full_prover, full_verifier, generate_setup_artifacts},
+    },
+    cryptocurrency::Cryptocurrency,
+    entry::Entry,
+    utils::{
+        amortized_kzg::{commit_kzg, create_naive_kzg_proof, verify_kzg_proof},
+        big_uint_to_fp, parse_csv_to_entries,
+    },
+};
+
+const K: u32 = 17;
+const N_CURRENCIES: usize = 2;
+const N_USERS: usize = 16;
+
+#[derive(Serialize, Deserialize)]
+struct CommitmentSolidityCallData {
+    range_check_snark_proof: String,
+    grand_sums_batch_proof: String,
+    total_balances: Vec<U256>,
+}
+
+#[derive(Serialize, Deserialize)]
+struct InclusionProofCallData {
+    proof: String,
+    challenges: Vec<U256>,
+    username: String,
+    user_values: Vec<U256>,
+}
+
+fn main() {
+    // Initialize with entries
+    let mut entries: Vec<Entry<N_CURRENCIES>> = vec![Entry::init_empty(); N_USERS];
+    let mut cryptos = vec![Cryptocurrency::init_empty(); N_CURRENCIES];
+
+    // Parse CSV to update entries and cryptos arrays
+    parse_csv_to_entries::<&str, N_CURRENCIES>("../csv/entry_16.csv", &mut entries, &mut cryptos)
+        .unwrap();
+
+    let univariate_grand_sum_circuit =
+        UnivariateGrandSum::<N_USERS, N_CURRENCIES>::init(entries.to_vec());
+
+    let (params, pk, _) = generate_setup_artifacts(
+        K,
+        Some("../backend/ptau/hermez-raw-17"),
+        &univariate_grand_sum_circuit,
+    )
+    .unwrap();
+
+    // Create a proof
+    let instances = vec![Fp::one(); 1]; // This instance is necessary to verify proof on solidity verifier.
+    let (zk_snark_proof, advice_polys, omega) = full_prover(
+        &params,
+        &pk,
+        univariate_grand_sum_circuit.clone(),
+        &[instances.clone()],
+    );
+
+    // Verify the proof to ensure validity
+    assert!(full_verifier(
+        &params,
+        pk.get_vk(),
+        &zk_snark_proof,
+        &[instances]
+    ));
+
+    let challenge = Fp::zero();
+    let mut csv_total: Vec<BigUint> = vec![BigUint::from(0u32); N_CURRENCIES];
+    for entry in &entries {
+        for (i, balance) in entry.balances().iter().enumerate() {
+            csv_total[i] += balance;
+        }
+    }
+
+    let poly_length = 1 << u64::from(K);
+    let total_balances = csv_total
+        .iter()
+        .map(|x| big_uint_to_fp(&(x)) * Fp::from(poly_length).invert().unwrap())
+        .collect::<Vec<Fp>>();
+
+    let mut grand_sums_kzg_proof = Vec::new();
+    for balance_column in 1..(N_CURRENCIES + 1) {
+        let f_poly = advice_polys.advice_polys.get(balance_column).unwrap();
+        let kzg_commitment = commit_kzg(&params, f_poly);
```

Comment:
This can be replaced by taking the committed point from the zk-proof transcript.

Reply:
https://github.com/summa-dev/summa-solvency/pull/268/files#diff-c68e26b4e20170b2d56ef3b1f9675c7b9ed5e2414278ec2edcd90f3fa1331474R99 here's the code that does that

Reply:
Great tip! It was much helped
It updated at https://github.com/summa-dev/summa-solvency/commit/6db5fb7cc56192741c86d55ff8f59c9eb7dea05b